doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1747,69 +1718,67 @@ name of the variable (here :g:`n`) is chosen based on :g:`T`.
              name the variable in the current goal and in the context of the
              existential variable. This can lead to surprising behaviors.
 
-.. tacv:: instantiate (@natural := @term)
-
-   This variant selects an existential variable by its position.  The
+   The second form refines an existential variable selected by its position.  The
    :n:`@natural` argument is the position of the existential variable
-   *from right to left* in the conclusion of the goal.  (Use one of
+   *from right to left* in the goal.  (Use one of
    the variants below to select an existential variable in a
    hypothesis.)  Counting starts at 1 and multiple occurrences of the
    same existential variable are counted multiple times.  Because this
    variant is not robust to slight changes in the goal, its use is
    strongly discouraged.
```

Comment:
I don't think we should be uniformly discouraging the use of this form.  As far as I'm aware, the ident := form is unusable in automation, because there's no way to get the name of an evar, and the identifier cannot be an Ltac variable holding the evar (I'd very much like to see this implemented...). The robust form for automation, when `ev` is an Ltac variable holding an evar, is `let H := fresh in pose ev as H; instantiate (1 := ltac:(tac)) in (value of H); clear H`.

Reply:
OK, I changed the last sentence to:

Note that this form is not robust to slight changes in the goal.

Reply:
Hm, I think it is fine to strongly discourage it *in most cases*.  Maybe " strongly discouraged except when you have very carefully set up the evars in the goal " or something?  It might also be worth explicitly mentioning this pattern, which is the robust way to use this variant

Reply:
@JasonGross Can your ltac statement be turned into an ltac-defined tactic (the `hloc` may be problematic)?  And if not, perhaps it would be good to define a new tactic or new variant that is robust.  Or perhaps we should add a way to get the name of an evar.  Why can't you use the first form?  Or should the first form also support `hloc`?  (But probably not something I would write the code for.)

Your equivalent jumps pretty deep into the details, could be a bit offputting to less experienced users.  Does your trick also works outside of ltac definitions?

> Maybe " strongly discouraged except when you have very carefully set up the evars in the goal " or something?

You only need the second form if you want hloc, right?

FWIW: perhaps it would be worthwhile if you came up with a list of Ltac tricks that could be better addressed by small improvements to tactics.  The best trick is no trick at all :-).

Reply:
> Your equivalent jumps pretty deep into the details, could be a bit offputting to less experienced users.

IMHO less experienced users should stay away from `instantiate`.

Reply:
I agree.  Maybe there should be some sort of way of marking some tactics, forms, or examples as "for advanced users" or "for tactic automation writers"?

Reply:
Well, we can do that in the text description. Note that we already mark some of them as "For internal use only."